docs/docs/aztec/writing_efficient_contracts.md

```diff
@@ -0,0 +1,358 @@
+
+:::note Difference with Rust
+Such designs with boolean flags lend themselves well into logical comparisons too since `&&` and `||` do not exist. With booleans, using `&` and `|` can give you the required logic efficiently. For more points specific to the Noir language, see [this](https://noir-lang.org/docs/explainers/explainer-writing-noir#translating-from-rust) section.
+
+:::
+
+### Optimization: Loop design
+
+Since private functions are circuits, their size must be known at compile time, which is equivalent to its execution trace.
+See [this example](https://github.com/noir-lang/noir-examples/blob/master/noir_by_example/loops/noir/src/main.nr#L11) for how to use loops when dynamic execution lengths (ie variable number of loops) is not possible.
+
+### Optimization: considered use of `unconstrained` functions
+#### Example - calculating square root
+
+Consider the following example of an implementation of the `sqrt` function:
+```rust
+use dep::aztec::macros::aztec;
+
+#[aztec]
+pub contract OptimisationExample {
+    use dep::aztec::macros::{functions::{initializer, private, public, utility}, storage::storage};
+
+    #[storage]
+    struct Storage<Context> {}
+
+    #[public]
+    #[initializer]
+    fn constructor() {}
+
+    #[private]
+    fn sqrt_inefficient(number: Field) -> Field {
+        super::sqrt_constrained(number)
+    }
+
+    #[private]
+    fn sqrt_efficient(number: Field) -> Field {
+        // Safety: calculate in unconstrained function, then constrain the result
+        let x = unsafe { super::sqrt_unconstrained(number) };
+        assert(x * x == number, "x*x should be number");
+        x
+    }
+
+}
+
+fn sqrt_constrained(number: Field) -> Field {
+    let MAX_LEN = 100;
+
+    let mut guess = number;
+    let mut guess_squared = guess * guess;
+    for _ in 1..MAX_LEN as u32 + 1 {
+        // only use square root part of circuit when required, otherwise use alternative part of circuit that does nothing
+        // Note: both parts of the circuit exist MAX_LEN times in the circuit, regardless of whether the square root part is used or not
+        if (guess_squared != number) {
+            guess = (guess + number / guess) / 2;
+            guess_squared = guess * guess;
+        }
+    }
+
+    guess
+}
+
+unconstrained fn sqrt_unconstrained(number: Field) -> Field {
+    let mut guess = number;
+    let mut guess_squared = guess * guess;
+    while guess_squared != number {
+        guess = (guess + number / guess) / 2;
+        guess_squared = guess * guess;
+    }
+    guess
+}
+
+```
+
+The two implementations after the contract differ in one being constrained vs unconstrained, as well as the loop implementation (which has other design considerations).
+Measuring the two, we find the `sqrt_inefficient` to require around 1500 extra gates compared to `sqrt_efficient`.
+
+To see each flamegraph:
+- `SERVE=1 aztec flamegraph target/optimisation_example-OptimisationExample.json sqrt_inefficient`
```

Comment:
probably remove this else we will need to manually maintain this if things change? defer to you?